File: web/src/lib/wire-line.ts
/**
 * THE TRAILING LINE A FEATURE SIGNS A MESSAGE BODY WITH, TAKEN OFF A PREVIEW — for every feature
 * that signs one, in ONE spelling.
 *
 * A game of chess and a companion are each carried by ordinary Teams messages whose last block is
 * `— <keyword> <6 lowercase hex> <payload>, via teams-lite`, read from the WORDS rather than from
 * markup. No surface that shows a message as a LINE OF TEXT may show that line: a chat row, a
 * channel row, and the page's own desktop notification all go through here, and the two features
 * used to hold a near-identical copy of the rule apiece. This module exists because that is where
 * the third feature's copy drifts — and it already had: the cut rules below were fixed in Rust and
 * on one of the two page copies, leaving the other showing a wire dump. It is the twin of
 * `push_policy::without_wire_line`, which serves a PUSH the same way and for the same reason (a
 * push has no page to strip on its behalf).
 *
 * **IT RE-VALIDATES THE TAIL BEFORE CUTTING**, never just cuts at the marker: a naive
 * `split(marker)[0]` truncates a real message that happens to contain the words. So the LAST
 * occurrence of the marker is found and then what follows it has to be a line — which is what
 * keeps an agent's own `— claude, via teams-lite` and a colleague's prose intact.
 *
 * **FOUR SHAPES COUNT AS ONE, because a preview is the body's first 120 characters** and a record
 * is one message rewritten on every act (`teams_read::preview_from_html`): a WHOLE line, one whose
 * PAYLOAD the cut broke, one whose ID the cut landed inside, and one where the cut landed inside the
 * MARKER ITSELF. Each feature's own measured crossing points are on its own strip; what matters here
 * is that all four arrive in practice and a rule holding only the first does nothing at all for a
 * record anybody has used.
 *
 * **THE FOURTH IS THE ONE THAT CANNOT BE FOUND BY LOOKING FOR THE MARKER**, which is why it is a
 * branch of its own: with the cut inside `— chess ` there is no marker in the text to find at all, so
 * `lastIndexOf` answers −1 and the fragment survives onto the row and into the push — the exact leak
 * the other three exist to stop. It is REACHABLE by arithmetic rather than hypothetical: the marker
 * occupies the code points immediately after the words, so where the id-cut rule needs words of
 * 105–110 characters (measured on chess), the marker-cut needs 111–118 — the very next window. An
 * earlier note here called the residual gap "one code point wide", which understated it by the whole
 * length of the marker.
 *
 * **THE THREE CUT RULES' PROOF IS WEAKER THAN THE WHOLE LINE'S, and every cost is nil-probability
 * rather than impossible.** What is left structurally is the marker, hex, and the cut. On the
 * PAYLOAD rule that is a six-character all-hex word, and English has those (`facade`, `decade`,
 * `deface`, `beaded`), so `"…— pet facade beats…"` is cut. On the ID rule it is **any hex-initial
 * word** — it fires at one character, so roughly 30% of English words by initial letter reach it
 * against roughly 0.5% for a full six-hex word, which is a widening of some 60x rather than of one
 * notch: `"…thanks ever — pet b…"` is cut, and so would `cage`, `bed`, `dish`, `food` and `bowl`
 * be, all of which a colleague could plainly write about a pet. The improbable part is the CUT
 * landing inside that 1–6 character window rather than the words. On the MARKER rule it is the em
 * dash alone, and that is the SMALLEST cost of the three rather than the widest: what it can take is
 * one dangling dash from a preview that really was cut (`"Hello there —…"`), never a word — see
 * {@link cutMarkerAt}. And an author's own trailing
 * ellipsis is indistinguishable from the preview's cut marker, so every one of the three fires on an
 * untruncated message that simply ends in one. None of them loses a message: the cost is a
 * trailing clause, which is what makes the trade acceptable.
 *
 * **CASE: LOWERCASE ONLY, in both languages.** A real id is `toString(16)`, so it can only ever be
 * lowercase — and Rust's `is_lower_hex` decides the same question the same way. Case-insensitive
 * here would therefore never admit a real ledger and only ever widen the prose paths above, and it
 * was unpinned in both directions until a test named it. Each feature's own `whole` pattern keeps
 * whatever case rule it already had: that grammar is the feature's, and it is also the branch prose
 * cannot reach.
 */

/** The PAYLOAD was cut: a whole id, a space, and then whatever fitted before the cut marker.
 *  LOWERCASE, matching Rust's `is_lower_hex` — see the case rule above. */
const PAYLOAD_CUT = /^[0-9a-f]{6}\s+.*…$/;

/** The ID ITSELF was cut: only hex, then the cut marker, and no payload at all. Measured on chess —
 *  words of 104–108 characters put the 120th code point inside the id, and 8 of 48 realistic
 *  engine-game ledgers landed there, leaking the marker fragment onto the row. LOWERCASE, for the
 *  reason above; the trailing space is unreachable from a real preview (`preview_from_html` trims
 *  before it appends the `…`) and is kept because it costs nothing and the rule is about the shape. */
const ID_CUT = /^[0-9a-f]{1,6} ?…$/;

/**
 * `text` with its trailing wire line removed, or `text` unchanged when the tail is not one.
 *
 * `marker` is the feature's own `— <keyword> ` — passed rather than merged, because which keyword a
 * body carries is exactly what tells the features apart. `whole` is that feature's own whole-line
 * pattern, which is where its version and payload grammar live.
 *
 * The MARKER-CUT rule is asked even when a marker IS present earlier in the text, because the two
 * anchor at opposite ends: `lastIndexOf` finds a marker anywhere, and a cut fragment is only ever the
 * last thing in the text. A body that says "— pet food is in the drawer" and is then cut inside its
 * own wire line holds both.
 */
export function withoutSignedLine(text: string, marker: string, whole: RegExp): string {
  const at = text.lastIndexOf(marker);
  if (at >= 0) {
    const tail = text.slice(at).trim();
    const rest = tail.slice(marker.length);
    if (whole.test(tail) || PAYLOAD_CUT.test(rest) || ID_CUT.test(rest)) return text.slice(0, at).trim();
  }
  const cut = cutMarkerAt(text, marker);
  return cut < 0 ? text : text.slice(0, cut).trim();
}

/**
 * Where a trailing fragment of the MARKER ITSELF starts, or −1 — `…23. Bxf6 — ches…`.
 *
 * The fragment must be a PROPER prefix of the marker: a whole one is the branch above, and every
 * shape under it. Longest first, so `— pet` is preferred over `—` and the cut takes the whole
 * fragment rather than leaving `pet` behind as a word.
 *
 * **THE SHORTEST PREFIX IS THE EM DASH ALONE, and stating what that costs is the point.** It fires on
 * any preview whose cut lands immediately after an em dash — with no keyword evidence at all — and it
 * is included because the em dash is genuinely where the cut can land (the marker's own first code
 * point). What it takes is that dangling em dash and nothing else: `"Hello there —…"` becomes
 * `"Hello there"`. That is a smaller cost than the id-cut rule stated above it, which can take a real
 * word, and it is the same trade — a trailing clause, never a message.
 */
function cutMarkerAt(text: string, marker: string): number {
  if (!text.endsWith("…")) return -1;
  const stem = text.slice(0, -1);
  for (let length = marker.length - 1; length > 0; length--) {
    if (stem.endsWith(marker.slice(0, length))) return stem.length - length;
  }
  return -1;
}

File: src/wire_line.rs
//! WHERE a feature's own signed line sits in a text, as the BACKEND reads one — for every feature
//! that signs a message body this way, in ONE spelling.
//!
//! A game of chess and a companion are each carried by ordinary Teams messages because Teams has no
//! private data channel: every challenge, move, spawn and feed is a message whose body carries one
//! machine-readable line — `— <keyword> <6 lowercase hex> <payload>, via teams-lite`. The whole shape
//! and every decision behind it live on the page ([`crate::chess_wire`] and `web/src/lib/pet-wire.ts`
//! say why); what this side answers is the ONE question a page cannot answer for a history it has not
//! loaded: **which of a conversation's stored messages carry a record at all.**
//!
//! **IT IS PARAMETERISED BY THE MARKER AND BY NOTHING ELSE, because the GRAMMAR is shared and the
//! keyword is not.** Both features spell the same `<6 lowercase hex> <anything>, via teams-lite` after
//! their own `— chess ` / `— pet `, so a second copy of that grammar per feature is a second chance to
//! narrow one of them and not the other — which is exactly how the third copy of the strip rules drifted
//! before `web/src/lib/wire-line.ts` was written (it is that module's twin, for a body rather than for a
//! preview).
//!
//! The grammar is deliberately narrow — six LOWERCASE hex characters and the exact trailing clause — so
//! an agent's own `— claude, via teams-lite` and a colleague's prose can never be read as a record. Case
//! matters: a real id is `toString(16)`, so it can only ever be lowercase, and admitting upper case
//! would only ever widen the prose paths.
//!
//! **WHAT IS NOT HERE: the STRIP.** Taking a line off a PREVIEW is `crate::push_policy`'s own, because a
//! preview is the body's first 120 characters and a record is one message rewritten on every act — so
//! that reader has to re-validate a tail the cut broke, in three shapes this one never sees. A body-shaped
//! reader and a preview-shaped reader are two questions, and merging them would give the wrong answer to
//! both.

/// Where the signed line starts, when this text ends with one — the byte offset of its em dash.
///
/// The text may be a message BODY (where the line sits inside `<p><em>…</em></p>`) or a flattened
/// preview (where it follows a newline), so a line runs to the next `<`, the next newline, or the end
/// of the text. Every occurrence is considered and the LAST qualifying one wins, which is what makes
/// this answer the same question for a body and for a preview — and what makes a body holding a QUOTE
/// of an earlier record resolve to the message's own line rather than to the one it quotes.
pub fn line_at(text: &str, marker: &str) -> Option<usize> {
    let mut found = None;
    let mut from = 0;
    while let Some(offset) = text[from..].find(marker) {
        let at = from + offset;
        let rest = &text[at + marker.len()..];
        let end = rest.find(['<', '\n']).unwrap_or(rest.len());
        if is_signed_line(rest[..end].trim()) {
            found = Some(at);
        }
        from = at + marker.len();
    }
    found
}

/// `<6 lowercase hex> <anything>, via teams-lite`, and nothing else counts.
fn is_signed_line(line: &str) -> bool {
    let Some((id, payload)) = line.split_once(' ') else {
        return false;
    };
    let is_id = id.len() == 6 && id.chars().all(|c| c.is_ascii_hexdigit() && !c.is_ascii_uppercase());
    is_id && payload.ends_with(", via teams-lite")
}

#[cfg(test)]
mod tests {
    use super::*;

    const CHESS: &str = crate::chess_wire::MARKER;
    const PET: &str = crate::pet_wire::MARKER;

    /// The "is there one at all" reading, spelled HERE rather than exported.
    ///
    /// Each feature already publishes its own (`carries_chess_line`, `carries_pet_line`) because that
    /// is what a store filter asks for by name; a third, marker-taking copy in this module would be a
    /// public function whose only caller is this test — the smell that had `chess_wire` carrying a
    /// diverged strip nothing called.
    fn carries(text: &str, marker: &str) -> bool {
        line_at(text, marker).is_some()
    }

    /// The line is read out of a message BODY, where it sits inside its own italic block — which is
    /// the shape a store filters on, and the one a preview-shaped reader would miss — and out of a
    /// flattened PREVIEW, where it follows a newline.
    #[test]
    fn a_line_is_found_in_a_body_and_in_a_preview() {
        let body = "<p>♟ Chess — I'd like a game.</p><p><em>— chess 7f3a1c v2 w open tc.600+0, via teams-lite</em></p>";
        assert!(carries(body, CHESS));
        assert!(carries("♟ 1. e4\n— chess 7f3a1c 1 e4, via teams-lite", CHESS));

        let pet = "<p>Nori is here.</p><p><em>— pet 7f3a1c v1 s.cat, via teams-lite</em></p>";
        assert!(carries(pet, PET));
        assert!(carries("Nori · fed 1\n— pet 7f3a1c v1 s.cat 1756060012345.f.7f3a1c, via teams-lite", PET));
    }

    /// EACH MARKER ANSWERS FOR ITS OWN FEATURE AND NOT FOR THE OTHER, which is the whole reason the
    /// keyword is the parameter: one store read must never count the other feature's messages.
    #[test]
    fn a_marker_answers_only_for_its_own_feature() {
        let chess = "— chess 7f3a1c v2 w open, via teams-lite";
        let pet = "— pet 7f3a1c v1 s.cat, via teams-lite";
        assert!(carries(chess, CHESS) && !carries(chess, PET));
        assert!(carries(pet, PET) && !carries(pet, CHESS));
    }

    /// And nothing else is ever read as a record: an agent's own signature, a colleague's prose, a
    /// bare em dash, an id that is not six hex characters, an UPPER-CASE one, and a line with the
    /// trailing clause missing.
    #[test]
    fn nothing_else_is_read_as_a_record() {
        for text in [
            "done — claude, via teams-lite",
            "on my way",
            "— chess not-a-game 1 e4, via teams-lite",
            "— chess 7F3A1C 1 e4, via teams-lite",
            "— chess 7f3a1c 1 e4",
            "— chess 7f3a1c",
        ] {
            assert!(!carries(text, CHESS), "{text} was read as a record");
        }
        for text in [
            "— pet food is in the second drawer, via teams-lite",
            "— pet 7F3A1C v1 s.cat, via teams-lite",
            "— pet 7f3a1c v1 s.cat",
        ] {
            assert!(!carries(text, PET), "{text} was read as a record");
        }
    }

    /// A body holding a QUOTE of an earlier record still carries its own line, and the one that
    /// counts is the message's own — the last of them.
    #[test]
    fn the_last_qualifying_line_is_the_one_that_counts() {
        let text = "— chess 7f3a1c 1 e4, via teams-lite\nand then — chess bbb222 2 e5, via teams-lite";
        let at = line_at(text, CHESS).expect("a line");
        assert_eq!(&text[at..], "— chess bbb222 2 e5, via teams-lite");
    }
}
